k8srm-prototype/pkg/api/capacity_types.go

```diff
@@ -38,19 +39,91 @@ type DevicePoolSpec struct {
 	// Attributes contains device attributes that are common to all devices
 	// in the pool.
 	// +optional
+	// +listType=atomic
 	Attributes []Attribute `json:"attributes,omitempty"`
 
-	// DeviceCount contains the total number of devices in the pool.
+	// SharedResources are pooled resources that are shared by all devices
+	// in the pool. This is typically used when representing a
+	// partitionable device, and need not be populated otherwise.
+	//
+	// +optional
+	// +listType=atomic
+	SharedResources []ResourceCapacity `json:"sharedResources,omitempty"`
+
+	// Devices contains the individual devices in the pool.
+	//
 	// +required
-	DeviceCount int `json:"count,omitempty"`
+	// +listType=atomic
+	Devices []Device `json:"devices,omitempty"`
```

Comment:
Carrying on from https://github.com/kubernetes-sigs/wg-device-management/pull/5/files#r1587245528 where the discussion had focused on modelling multiple devices in a pool vs a single device. 

> Do they put all of their simple devices in a single DevicePool alongside all partitionable devices in their own individual DevicePools?

That's the ambiguity I dislike, but maybe it's not such a big deal.  Sometimes it means N devices and sometimes in means 1"macrodevice".

> Why is there a "pool" boundary at all if there is no real meaning tied to it?
> is it only there so as to avoid having lots of separate Device objects

There are 3 things a pool does, IMO:
1) Shared attributes (could just be denormalized)
2) Shared resources (gives the mutual-exclusion between MIG modes)
3) Avoid big-N Device objects (I suspect N won't be big per-node)

 Calling it "pool" implies a sameness that isn't required or enforced.  Though experiment - what if it really was called Device.  If you have 2 simple devices you would model it as 2 Device objects.  If you have 2 MIG-ish devices, you model it as 2 Devices with sub-devices.

If you had 5K nodes and each had 8 MIG-ish devices, each with 15 subdevices, that would be 40K Device objects and 600K sub-device structures.  Annoying but possible to dedup in app-memory.  These should not change frequently, right?  They have matching status Conditions - how are we thinking about that?  Also, these resources won't have a lot of watches, right?

If we do pool them, the condition updates could be batched, at the cost of latency, but only for simple devices, since MIG-ish ones would be one-per-pool anyway.

We should think about scalability, but only AFTER we pin down utility.

@johnbelamaric thoughts?  How often are we ACTUALLY going to have more than one device in a Pool?  How would we model networking devices here (including multi-net and software-defined devices like VXLAN?

Reply:
Yeah, I actually started to wonder the same thing. I put it in a comment somewhere but I can't find it now. I started looking at what a multi-PF SR-IOV network card might look like in the capacity model, and it sured looked a lot like the partitionable devices.

The scheduler code will have to flatten the pools anyway. So if we are comfortable with the scaling factors of not having any denormalization in pools, then we can just have Device objects. But...are those two different top-level objects? Device and PartitionableDevice or something? I am struggling to see what that model looks like. Is it just the "Pool" in the name that is confusing? Do we go back to Slice?

Reply:
Found that [comment](https://github.com/kubernetes-sigs/wg-device-management/pull/5/files#discussion_r1590004037), it's in the thread you already linked.